src/test/kotlin/no/nav/syfo/application/service/ForhandsvarselServiceSpek.kt

```diff
@@ -0,0 +1,128 @@
+package no.nav.syfo.application.service
+
+import io.ktor.server.testing.*
+import io.mockk.*
+import kotlinx.coroutines.runBlocking
+import no.nav.syfo.UserConstants.ARBEIDSTAKER_PERSONIDENT
+import no.nav.syfo.UserConstants.PDF_FORHANDSVARSEL
+import no.nav.syfo.UserConstants.VEILEDER_IDENT
+import no.nav.syfo.application.IVurderingRepository
+import no.nav.syfo.generator.generateDocumentComponent
+import org.amshove.kluent.internal.assertFailsWith
+import org.amshove.kluent.shouldBeEqualTo
+import org.spekframework.spek2.Spek
+import org.spekframework.spek2.style.specification.describe
+import java.lang.RuntimeException
+
+object ForhandsvarselServiceSpek : Spek({
+
+    describe(ForhandsvarselServiceSpek::class.java.simpleName) {
+        with(TestApplicationEngine()) {
+            start()
+            val vurderingRepositoryMock = mockk<IVurderingRepository>(relaxed = true)
+            val varselPdfServiceMock = mockk<VarselPdfService>(relaxed = true)
+            val forhandsvarselService = ForhandsvarselService(
+                vurderingRepository = vurderingRepositoryMock,
+                varselPdfService = varselPdfServiceMock,
+            )
+
+            beforeEachTest {
+                clearAllMocks()
+                coEvery {
+                    vurderingRepositoryMock.createForhandsvarsel(any(), any(), any())
+                } returns Unit
+                coEvery {
+                    varselPdfServiceMock.createVarselPdf(any(), any(), any())
+                } returns PDF_FORHANDSVARSEL
+            }
+
+            describe("Forhåndsvarsel") {
+                val begrunnelse = "En begrunnelse"
+                val document = generateDocumentComponent(begrunnelse)
+                describe("Happy path") {
+                    it("Creates forhåndsvarsel") {
```

Comment:
Jeg synes man godt kan teste med "ekte" repository og pdf-service også, slik at man får testet at det man forventer blir lagret i databasen og vi får laget en pdf (ut fra en pdfgen-mock). Eller tenker du gjøre det med en api-test senere?

Reply:
Jeg synes at api-tester eller tilsvarende som tester mer integrasjoner og med mindre mocks stort sett gir mer verdi og er det viktigte å ha på plass. Så kan vi alltids supplere med mer rene enhetstester på servicen, kanskje spesielt hvis det er mer logikk der man ønsker å teste spesielt.

Reply:
Ja, tanken var vel å teste servicen her, og at APIet tester "alt" 🤔 Jeg er veldig åpen for begge deler, så lenge vi har _noe_ integrasjonstester, som du sier 👍🏼

Men jeg vet ikke om vi ønsker å gå for API-test på følgende måte:
1. Gjøre POST og så sjekke at det gikk bra med en GET, eller
2. Gjøre POST og så sjekke rett i databasen

Tradisjonelt sett har vi vel gjort mest av 2, men det handler kanskje bare om hvor mye vi ønsker å blande inn i testene av avhengigheter. Hvis vi er veldig religiøse på det så er vel nr 1 best for da forholder man seg kun til API-kall i API-testen. Samtidig kan man gi unntak fra sånne strenge regler i tester hvis vi føler "målet helliger middelet" 🤷🏼‍♂️

Reply:
Jeg synes egentlig det er ryddig å teste `POST` og `GET` mer isolert for seg - dvs gjøre `POST` og sjekke databasen og egen test på `GET` som først oppretter noe i databasen. Men det blir jo litt mer testdata-oppsett av det :man_shrugging:

Reply:
Hadde tenkt at det kan være bra å et sett med integrasjonstester / tester mot api laget, og unit tester for de andre lagene. Så isolerte tester for lagring/repositories, tester for applikasjonslaget om vi trenger, og i alle fall tester for domene laget.
Litt av målet (slik jeg har forstått det) med clean architecture er at man også skal kunne teste forretningslogikk helt isolert fra lagring og andre tredjeparter.